src/symbol/collision_index.js

```diff
@@ -0,0 +1,353 @@
+// @flow
+
+const Point = require('@mapbox/point-geometry');
+const intersectionTests = require('../util/intersection_tests');
+
+const Grid = require('./grid_index');
+const glmatrix = require('@mapbox/gl-matrix');
+
+const mat4 = glmatrix.mat4;
+
+const projection = require('../symbol/projection');
+
+import type Transform from '../geo/transform';
+import type TileCoord from '../source/tile_coord';
+import type {SingleCollisionBox} from '../data/bucket/symbol_bucket';
+
+// When a symbol crosses the edge that causes it to be included in
+// collision detection, it will cause changes in the symbols around
+// it. This constant specifies how many pixels to pad the edge of
+// the viewport for collision detection so that the bulk of the changes
+// occur offscreen. Making this constant greater increases label
+// stability, but it's expensive.
+const viewportPadding = 100;
+
+/**
+ * A collision index used to prevent symbols from overlapping. It keep tracks of
+ * where previous symbols have been placed and is used to check if a new
+ * symbol overlaps with any previously added symbols.
+ *
+ * There are two steps to insertion: first placeCollisionBox/Circles checks if
+ * there's room for a symbol, then insertCollisionBox/Circles actually puts the
+ * symbol in the index. The two step process allows paired symbols to be inserted
+ * together even if they overlap.
+ *
+ * @private
+ */
+class CollisionIndex {
+    grid: Grid;
+    ignoredGrid: Grid;
+    transform: Transform;
+    pitchfactor: number;
+
+    constructor(
+        transform: Transform,
+        grid: Grid = new Grid(transform.width + 2 * viewportPadding, transform.height + 2 * viewportPadding, 25),
+        ignoredGrid: Grid = new Grid(transform.width + 2 * viewportPadding, transform.height + 2 * viewportPadding, 25)
+    ) {
+        this.transform = transform;
+
+        this.grid = grid;
+        this.ignoredGrid = ignoredGrid;
+        this.pitchfactor = Math.cos(transform._pitch) * transform.cameraToCenterDistance;
+    }
+
+    placeCollisionBox(collisionBox: SingleCollisionBox, allowOverlap: boolean, textPixelRatio: number, posMatrix: mat4): Array<number> {
```

Comment:
Yeah, that name is left over from the previous behavior. How about instead of "placeX" which returns true if the item _can_ be placed, we name the functions "collisionTextX", and return true if there's a collision?